src/kern/npf_alg_pptp.c

```diff
@@ -0,0 +1,788 @@
+}
+
+/*
+ * PPTP GRE ALG translator.
+ */
+static bool
+npfa_pptp_gre_translate(npf_cache_t *npc, npf_nat_t *nt, bool forw)
+{
+	nbuf_t *nbuf = npc->npc_nbuf;
+	pptp_gre_hdr_t *gre;
+	pptp_gre_con_slot_t gre_slot;
+
+	if (forw || !npf_iscached(npc, NPC_IP4 | NPC_ALG_PPTP_GRE) ||
+			  npc->npc_proto != IPPROTO_GRE)
+		return false;
+
+	nbuf_reset(nbuf);
+	gre = nbuf_advance(nbuf, npc->npc_hlen, sizeof(pptp_gre_hdr_t));
+	if (gre == NULL)
+		return false;
+
+	gre_slot.u64 = (uint64_t)npf_nat_get_alg_arg(nt);
+	KASSERT(gre->call_id == gre_slot.ctx.client_call_id);
+	gre->call_id = gre_slot.orig_client_call_id;
+
+	return true;
+}
+
+/*
+ * Destroy gre connection.
+ * Put translated call-id back to the portmap.
+ */
+static void
+npfa_pptp_gre_destroy(npf_t *npf, npf_conn_t *con)
+{
+	pptp_gre_con_slot_t gre_slot;
+	npf_nat_t *nt;
+	npf_connkey_t *fw;
+	npf_addr_t ips[2];
+	uint16_t ids[2];
+	uint16_t alen;
+	uint16_t proto;
+
+	fw = npf_conn_getforwkey(con);
+	nt = con->c_nat;
+
+	/* only ipv4 is supported */
+	if (NPF_CONNKEY_ALEN(fw) != sizeof(uint32_t) || nt == NULL)
+		return;
+
+	npf_connkey_getkey(fw, &proto, ips, ids, &alen);
+
+	gre_slot.u64 = (uint64_t)npf_nat_get_alg_arg(nt);
+	if (gre_slot.ctx.client_call_id != 0)
+		npfa_translated_call_id_put(&ips[NPF_DST], gre_slot.ctx.client_call_id);
+}
+
+/*
+ * npf_alg_icmp_{init,fini,modcmd}: ICMP ALG initialization, destruction
+ * and module interface.
+ */
+
+__dso_public int
+npf_alg_pptp_init(npf_t *npf)
+{
+	static const npfa_funcs_t pptp_tcp = {
+		.match     = npfa_pptp_tcp_match,
+		.translate = npfa_pptp_tcp_translate,
+		.inspect   = NULL,
+		.destroy   = npfa_pptp_tcp_destroy,
+	};
+
+	static const npfa_funcs_t pptp_gre = {
+		.match     = NULL,
+		.translate = npfa_pptp_gre_translate,
+		.inspect   = NULL,
+		.destroy   = npfa_pptp_gre_destroy,
+	};
+
+	/* call_id range */
+	pptp_alg.pm_params.min_port = 1;
+	pptp_alg.pm_params.max_port = UINT16_MAX;
+
+	pptp_alg.pm = npf_portmap_init_pm();
+	if (pptp_alg.pm == NULL)
+		return ENOMEM;
+
+	pptp_alg.alg_pptp_tcp = npf_alg_register(npf, "pptp_tcp", &pptp_tcp);
+	if (pptp_alg.alg_pptp_tcp == NULL)
+		return ENOMEM;
+
+	pptp_alg.alg_pptp_gre = npf_alg_register(npf, "pptp_gre", &pptp_gre);
+	if (pptp_alg.alg_pptp_tcp == NULL) {
+		npf_alg_unregister(npf, pptp_alg.alg_pptp_tcp);
+		return ENOMEM;
+	} else {
+		return 0;
+	}
+}
+
+__dso_public int
+npf_alg_pptp_fini(npf_t *npf)
+{
+	KASSERT(pptp_alg.alg_pptp_tcp != NULL);
+	KASSERT(pptp_alg.alg_pptp_gre != NULL);
+	npf_portmap_fini_pm(pptp_alg.pm);
+	npf_alg_unregister(npf, pptp_alg.alg_pptp_tcp);
+	return npf_alg_unregister(npf, pptp_alg.alg_pptp_gre);
+}
+
+void
+npf_pptp_conn_conkey(const npf_cache_t *npc, uint16_t *id, bool forw)
+{
+	const pptp_gre_context_t *pptp_gre_ctx;
+	const pptp_gre_hdr_t *pptp_gre_hdr;
+
+	KASSERT(npf_iscached(npc, NPC_ALG_PPTP_GRE | NPC_ALG_PPTP_GRE_CTX));
+	if (npf_iscached(npc, NPC_ALG_PPTP_GRE_CTX)) {
+		pptp_gre_ctx = npc->npc_l4.hdr;
+		if (forw) {
```

Comment:
Just an idea: if `pptp_gre_context_t` would have `uint16_t call_id[2];` i.e. an array, then you could just re-use `NPF_SRC`/`NPF_DST` (or define PPTP_{CLIENT,SERVER}_CALL_ID and CTASSERT() them to be equal to NPF_SRC/NPF_DST) simplify some handling here and there.

Reply:
I don't think it's worth it.